### DISCUSSION TITLE
Missing functions in `FloatingPoint` API?

### DISCUSSION THREAD

**Discussion by RyanGlScott** [B-Lang-org/bsc#534] [General]
I've been playing around with the [`FloatingPoint`](https://github.com/B-Lang-org/bsc/blob/0b4f8928ecdd67a3b08dca26917e660091eb64f6/src/Libraries/Base3-Math/FloatingPoint.bsv) package for a bit, and while it mostly does what I want, I was surprised not to find versions of the standard trigonometric functions (`sin`, `cos`, `tan`, etc.) for `FloatingPoint`. Is their omission intentional? I noticed that they _are_ [defined over `Real`s](https://github.com/B-Lang-org/bsc/blob/26e119fb3ecfbbc1ec6b676e7cf26646d48f5eec/src/Libraries/Base1/Real.bs#L3-L6), but `Real` doesn't suffice for my needs since it is [not an instance of `Bits`](https://github.com/B-Lang-org/bsc/issues/368).

**Comment by quark17** [comment]
I don't think there's a reason.  The users of the `FloatingPoint` package have often been people writing processors, so the functions that exist are the ones typically needed for that, and sin/cos/etc haven't been ops in those ISAs.  Feel free to contribute implementations for other functions.  I'm not experienced in hardware for floating point, but I'd guess that there are many possible implementations, so one library isn't going to fit everyone's needs, and I suspect that most users will want to write their own or import IP (perhaps using native constructs in the target hardware) or so on.  But there's no reason why BSC couldn't have a simple implementation available in its libraries, for incremental design or for people who don't have tight resource requirements.  BSC also isn't going to be able to promise that the implementation that it does provide is optimal in any way.  But with those caveats, it's not unreasonable to add implementations, if someone has written them and wants to submit them.  There is also a repo `bsc-contrib` where libraries can be made available.  There's perhaps a debate to be had about how much should be in the standard libraries vs what eventually should be considered extra, that can people can download and use if they need it.

**Reply by RyanGlScott** [reply]
Ah, thanks for the background context.

I don't have my own implementations to offer, but I'll keep `bsc-contrib` in mind for the future.

**Reply by srd8668** [reply]
Hi i want to use the floating point package to develop an accelerator, where can i find the documentation for it to know how to use it.

**Reply by quark17** [reply]
@srd8668 , I have responded to that question on Discussion #709 .

**Comment by RyanGlScott** [comment]
Some other functions from the `Real` package that would be nice to have `FloatingPoint` counterparts for:

* Proper implementations of [`(**)`, `exp_e`, `log`, and `logb`](https://github.com/B-Lang-org/bsc/blob/0b4f8928ecdd67a3b08dca26917e660091eb64f6/src/Libraries/Base3-Math/FloatingPoint.bsv#L1049-L1067). These are part of `FloatingPoint`'s `Arith` instance, but they simply call `error` when used. Unfortunately, I don't really know how to implement these myself.
* [`ceil` and `floor`](https://github.com/B-Lang-org/bsc/blob/26e119fb3ecfbbc1ec6b676e7cf26646d48f5eec/src/Libraries/Base1/Real.bs#L87-L91) functions. Interestingly, there is an unexported [`round`](https://github.com/B-Lang-org/bsc/blob/0b4f8928ecdd67a3b08dca26917e660091eb64f6/src/Libraries/Base3-Math/FloatingPoint.bsv#L560-L650) function in `FloatingPoint.bsv` that could be used to implement `ceil` and `floor`. Something like this would suffice:

  ```bluespec
  ceil :: FloatingPoint e m -> FloatingPoint e m
  ceil x = round Rnd_Plus_Inf x ???

  floor :: FloatingPoint e m -> FloatingPoint e m
  floor x = round Rnd_Minus_Inf x ???
  ```

  I'm not really sure what would go in place of the `???`s, however, as I don't understand what purpose the `guard` argument to `round` is serving.

**Comment by RyanGlScott** [comment]
One possible starting point for a native Bluespec implementation is to adapt the code in the [LibBF](https://bellard.org/libbf/) library, which provides an implementation of arbitrary-precision IEEE754 floating-point numbers in a style very similar to what Bluespec's `FloatingPoint` package is using. Granted, LibBF is a software implementation, not a hardware one, but it should still be useful as a starting point.
